README
NAME

    Class::XSAccessor - Generate fast XS accessors without runtime
    compilation

SYNOPSIS

      package MyClass;
      use Class::XSAccessor
        replace     => 1,   # Replace existing methods (if any)
        constructor => 'new',
        getters     => {
          get_foo => 'foo', # 'foo' is the hash key to access
          get_bar => 'bar',
        },
        setters => {
          set_foo => 'foo',
          set_bar => 'bar',
        },
        accessors => {
          foo => 'foo',
          bar => 'bar',
        },
        predicates => {
          has_foo => 'foo',
          has_bar => 'bar',
        },
        lvalue_accessors => { # see below
          baz => 'baz', # ...
        },
        true  => [ 'is_token', 'is_whitespace' ],
        false => [ 'significant' ];

      # The imported methods are implemented in fast XS.

      # normal class code here.

    As of version 1.05, some alternative syntax forms are available:

      package MyClass;

      # Options can be passed as a HASH reference, if preferred,
      # which can also help Perl::Tidy to format the statement correctly.
      use Class::XSAccessor {
         # If the name => key values are always identical,
         # the following shorthand can be used.
         accessors => [ 'foo', 'bar' ],
      };

DESCRIPTION

    Class::XSAccessor implements fast read, write and read/write accessors
    in XS. Additionally, it can provide predicates such as "has_foo()" for
    testing whether the attribute "foo" is defined in the object. It only
    works with objects that are implemented as ordinary hashes.
    Class::XSAccessor::Array implements the same interface for objects that
    use arrays for their internal representation.

    Since version 0.10, the module can also generate simple constructors
    (implemented in XS). Simply supply the "constructor =>
    'constructor_name'" option or the "constructors => ['new', 'create',
    'spawn']" option. These constructors do the equivalent of the following
    Perl code:

      sub new {
        my $class = shift;
        return bless { @_ }, ref($class)||$class;
      }

    That means they can be called on objects and classes but will not clone
    objects entirely. Parameters to "new()" are added to the object.

    The XS accessor methods are between 3 and 4 times faster than typical
    pure-Perl accessors in some simple benchmarking. The lower factor
    applies to the potentially slightly obscure "sub set_foo_pp
    {$_[0]->{foo} = $_[1]}", so if you usually write clear code, a factor of
    3.5 speed-up is a good estimate. If in doubt, do your own benchmarking!

    The method names may be fully qualified. The example in the synopsis
    could have been written as "MyClass::get_foo" instead of "get_foo". This
    way, methods can be installed in classes other than the current class.
    See also: the "class" option below.

    By default, the setters return the new value that was set, and the
    accessors (mutators) do the same. This behaviour can be changed with the
    "chained" option - see below. The predicates return a boolean.

    Since version 1.01, "Class::XSAccessor" can generate extremely simple
    methods which just return true or false (and always do so). If that
    seems like a really superfluous thing to you, then consider a large
    class hierarchy with interfaces such as PPI. These methods are provided
    by the "true" and "false" options - see the synopsis.

OPTIONS

    In addition to specifying the types and names of accessors, additional
    options can be supplied which modify behaviour. The options are
    specified as key/value pairs in the same manner as the accessor
    declaration. For example:

      use Class::XSAccessor
        getters => {
          get_foo => 'foo',
        },
        replace => 1;

    The list of available options is:

  replace
    Set this to a true value to prevent "Class::XSAccessor" from complaining
    about replacing existing subroutines.

  chained
    Set this to a true value to change the return value of setters and
    mutators (when called with an argument). If "chained" is enabled, the
    setters and accessors/mutators will return the object. Mutators called
    without an argument still return the value of the associated attribute.

    As with the other options, "chained" affects all methods generated in
    the same "use Class::XSAccessor ..." statement.

  class
    By default, the accessors are generated in the calling class. The the
    "class" option allows the target class to be specified.

LVALUES

    Support for lvalue accessors via the keyword "lvalue_accessors" was
    added in version 1.08. At this point, THEY ARE CONSIDERED HIGHLY
    EXPERIMENTAL. Furthermore, their performance hasn't been benchmarked
    yet.

    The following example demonstrates an lvalue accessor:

      package Address;
      use Class::XSAccessor
        constructor => 'new',
        lvalue_accessors => { zip_code => 'zip' };

      package main;
      my $address = Address->new(zip => 2);
      print $address->zip_code, "\n"; # prints 2
      $address->zip_code = 76135; # <--- This is it!
      print $address->zip_code, "\n"; # prints 76135

CAVEATS

    Probably won't work for objects based on *tied* hashes. But that's a
    strange thing to do anyway.

    Scary code exploiting strange XS features.

    If you think writing an accessor in XS should be a laughably simple
    exercise, then please contemplate how you could instantiate a new XS
    accessor for a new hash key that's only known at run-time. Note that
    compiling C code at run-time a la Inline::C is a no go.

    Threading. With version 1.00, a memory leak has been fixed. Previously,
    a small amount of memory would leak if "Class::XSAccessor"-based classes
    were loaded in a subthread without having been loaded in the "main"
    thread. If the subthread then terminated, a hash key and an int per
    associated method used to be lost. Note that this mattered only if
    classes were only loaded in a sort of throw-away thread.

    In the new implementation, as of 1.00, the memory will still not be
    released, in the same situation, but it will be recycled when the same
    class, or a similar class, is loaded again in any thread.

SEE ALSO

    *   Class::XSAccessor::Array
    *   AutoXS

AUTHORS

    Steffen Mueller <smueller@cpan.org>
    chocolateboy <chocolate@cpan.org>

COPYRIGHT AND LICENSE

    Copyright (C) 2008-2020 by Steffen Mueller

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8 or, at your
    option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Class-XSAccessor.

1.20 - TBD

  - B::C compatibility fix (thanks, Todd Rinaldo)
  - Clean up macros (thanks, Nicolas R)
  - Re-enable erroneously disabled test (thanks, Peter Rabbitson)

1.19 - 2013-11-22
  - Require Time::HiRes explicitly because CentOS cripples the base
    perl install.
  - Simplify entersub optimization (chocolateboy)

1.18 - 2013-06-17
  - Revert fixes for implicitly-lvalue getters for now since that
    actually breaks user code. It seems it's not just because the
    users are naughty, so more investigation required.

1.17 - 2013-06-17
  - For Hashes: Implement predicates that check definedness and
    existance explicitly. The traditional "predicates" check
    definedness. It's conceivable to also want a bool-check
    type predicate. File a ticket if you need that.
  - Fix bug regarding getters being implicitly lvalue by
    returning the internal SV*. Instead, we now use TARG.

1.16 - 2012-11-05
  - Drop erroneous MYMETA files from distribution.

1.15 - 2012-11-04
  - Support for hash keys with NUL (\0) characters. Previously,
    these were truncated.

1.14 - 2012-08-26
  - Skip some failing tests on old debugging perls. Guys, please
    upgrade your perl!

1.13 - 2011-12-12
  - Promotes 1.12_03 to a stable release.

1.12_03 - 2011-12-09
  - Fix for unthreaded perls (broken in 1.12_02).

1.12_02 - 2011-12-07
  - Removes the cached read-only and read-write accessors
    for the time being. (These were only available from another
    development releasse.
  - Much more brutal thread-safety testing.
  - Fixed thread-safety problem with the global hashkey storage.
  - Lots of refactoring in the C code.
  - Instead of storing an index in the CV, we store a pointer to
    the hashkey struct.

1.12_01 - 2011-11-29
  - Implements cached read-only and read-write accessors.
    Details on what that means are in the documentation.

1.12 - 2011-09-04
  - Reclaim compatibility with the most recent versions of
    ExtUtils::ParseXS.
  - Explicit tests for wrong-type invocants.

1.11 - 2010-12-03
  - Fix assignment to lvalue accessors that
    point at an uninitialized hash element.

1.10 - 2010-12-01
  - Fix RT #63458 and potentially #50454
    We don't occasionally crash during END any more.
    Instead, we rely on the OS to reap a bit of memory after
    perl was shut down anyway.
  - Tiny refactoring for smaller object size.

1.09 - 2010-10-31
  - Fix #62531: Predicates return value, not bool (SJOHNSTON)
  - TODO test for perl-crashing bug (in perl) that can happen
    on (arcane) XSUB aliasing on perls < 5.8.9 (Peter Rabbitson)
    We're open for work-around patches.

1.08 - 2010-09-17
  - Promote latest development release to a stable release.

1.07_04 - 2010-09-12
  - Since WIN32 doesn't have the PERL_CORE optimization,
    it gets the PERL_NO_GET_CONTEXT optimization back.
  - Add threading test that would previously crash on win32
    and perls compiled with track-mempool.
  - Use the system's malloc/etc for the shared memory, not perl's.

1.07_03 - 2010-09-09
  - Minor constructor optimization/cleanup.
  - Various built-time warning fixes.
  - PERL_CORE optimization now disabled on WIN32.
  - Class::Accessor::Fast compatibility code added (not
    for public consumption!)
  - Clear requirement of Perl 5.8 everywhere.
  - Fix minor (constant as in O(1)) memory leak.

1.07_02 - 2010-08-23
  - Various warning fixes and small cleanups over previous
    dev. version.

1.07_01 - 2010-08-18
  - Experimental support for lvalue accessors:
    $obj->foo = 12

1.07 - 2010-08-15
  - Include two new test files for the fix in 1.06.
  - Define PERL_CORE, but *only* while including XSUB.h to get
    a significant speed-up (see XSAccessor.xs for an explanation).
    Idea from chocolateboy. Complaints from rightfully annoyed
    perl5-porters (in particular but not limited to Nicholas)
    go to Steffen.

1.06 - 2010-08-14
  - Add sanity checks to make sure we don't segfault on
    invalid invocants (chocolateboy)

1.05 - 2009-11-15
  - Minor developer doc tweaks.
  - Minor XS refactoring

1.04_05 - 2009-11-09
  - Fixes for perls < 5.10:
    => No entersub optimization
    => Do no use precalculated hashes
  - Updated entersub optimization
  - Remove brain-damaged double-hashing
  - Minor portability fixlets

1.04_04 - 2009-11-05
  - Fixes for non-threaded perls
    (no need for locks, perl_mutex not even defined).

1.04_03 - 2009-11-03
  ** This release features some very radical changes. Test well. **
  - Replace use of perl hashes in the global hash key name storage with
    a full-blown, separate implementation of a hash table
    (Steffen, chocolateboy)
  - Similarly, throw out the SV's for simple C strings.
  - Add a global lock for all modifications to global data structures:
  - The above three items fix RT #50454 (serious threading issues).
  - Add support for alternate use Class::XSAccessor { ... } syntax
    (Adam K)

1.04_02 - 2009-09-07
  ** This release features some very radical changes. Test well. **
  - Significant optimization by replacing the relevant entersub ops
    with stripped down versions (chocolateboy)

1.04_01 - 2009-09-07
  ** This release features some very radical changes. Test well. **
  - More aggressive OPTIMIZE flags if possible (chocolateboy)
  - Added shorthand syntax for getters, setters, accessors, and predicates
    where the attribute has the same name as the method (chocolateboy)
  - Remove dependency on AutoXS::Header.
  - Merge Class::XSAccessor::Array into this distribution.
  - Refactored the XS to remove duplicate code.
  - Refactored the perl code in XSAccessor.pm and Array.pm to remove
    duplicate code (=> Heavy.pm).
  - Upgrade Devel::PPPort/ppport.h (chocolateboy)

1.04 - 2009-06-11
  - Fix a bunch of warnings thanks to a heads up from
    Marcela Maslanova.

1.03 - 2009-05-31
  - Upgrade to AutoXS::Header 1.01: Make it C89!

1.02 - 2009-05-17

lib/Class/XSAccessor.pm
package Class::XSAccessor;
use 5.008;
use strict;
use warnings;
use Carp qw/croak/;
use Class::XSAccessor::Heavy;
use XSLoader;

our $VERSION = '1.19';

XSLoader::load('Class::XSAccessor', $VERSION);

sub _make_hash {
  my $ref = shift;

  if (ref ($ref)) {
    if (ref($ref) eq 'ARRAY') {
      $ref = { map { $_ => $_ } @$ref }
    } 
  } else {
    $ref = { $ref, $ref };
  }

  return $ref;
}

sub import {
  my $own_class = shift;
  my ($caller_pkg) = caller();

  # Support both { getters => ... } and plain getters => ...
  my %opts = ref($_[0]) eq 'HASH' ? %{$_[0]} : @_;

  $caller_pkg = $opts{class} if defined $opts{class};

  # TODO: Refactor. Move more duplicated code to ::Heavy
  my $read_subs      = _make_hash($opts{getters} || {});
  my $set_subs       = _make_hash($opts{setters} || {});
  my $acc_subs       = _make_hash($opts{accessors} || {});
  my $lvacc_subs     = _make_hash($opts{lvalue_accessors} || {});
  my $pred_subs      = _make_hash($opts{predicates} || {});
  my $ex_pred_subs   = _make_hash($opts{exists_predicates} || {});
  my $def_pred_subs  = _make_hash($opts{defined_predicates} || {});
  my $test_subs      = _make_hash($opts{__tests__} || {});
  my $construct_subs = $opts{constructors} || [defined($opts{constructor}) ? $opts{constructor} : ()];
  my $true_subs      = $opts{true} || [];
  my $false_subs     = $opts{false} || [];

  foreach my $subtype ( ["getter", $read_subs],
                        ["setter", $set_subs],
                        ["accessor", $acc_subs],
                        ["lvalue_accessor", $lvacc_subs],
                        ["test", $test_subs],
                        ["ex_predicate", $ex_pred_subs],
                        ["def_predicate", $def_pred_subs],
                        ["def_predicate", $pred_subs] )
  {
    my $subs = $subtype->[1];
    foreach my $subname (keys %$subs) {
      my $hashkey = $subs->{$subname};
      _generate_method($caller_pkg, $subname, $hashkey, \%opts, $subtype->[0]);
    }
  }

  foreach my $subtype ( ["constructor", $construct_subs],
                        ["true", $true_subs],
                        ["false", $false_subs] )
  {
    foreach my $subname (@{$subtype->[1]}) {
      _generate_method($caller_pkg, $subname, "", \%opts, $subtype->[0]);
    }
  }
}

sub _generate_method {
  my ($caller_pkg, $subname, $hashkey, $opts, $type) = @_;

  croak("Cannot use undef as a hash key for generating an XS $type accessor. (Sub: $subname)")
    if not defined $hashkey;

  $subname = "${caller_pkg}::$subname" if $subname !~ /::/;

  Class::XSAccessor::Heavy::check_sub_existence($subname) if not $opts->{replace};
  no warnings 'redefine'; # don't warn about an explicitly requested redefine

  if ($type eq 'getter') {
    newxs_getter($subname, $hashkey);
  }
  elsif ($type eq 'lvalue_accessor') {
    newxs_lvalue_accessor($subname, $hashkey);
  }
  elsif ($type eq 'setter') {
    newxs_setter($subname, $hashkey, $opts->{chained}||0);
  }
  elsif ($type eq 'def_predicate') {
    newxs_defined_predicate($subname, $hashkey);
  }
  elsif ($type eq 'ex_predicate') {
    newxs_exists_predicate($subname, $hashkey);
  }
  elsif ($type eq 'constructor') {
    newxs_constructor($subname);
  }
  elsif ($type eq 'true') {
    newxs_boolean($subname, 1);
  }
  elsif ($type eq 'false') {
    newxs_boolean($subname, 0);
  }
  elsif ($type eq 'test') {
    newxs_test($subname, $hashkey);
  }
  else {
    newxs_accessor($subname, $hashkey, $opts->{chained}||0);
  }
}

1;

__END__

=head1 NAME

Class::XSAccessor - Generate fast XS accessors without runtime compilation

=head1 SYNOPSIS

  package MyClass;
  use Class::XSAccessor
    replace     => 1,   # Replace existing methods (if any)
    constructor => 'new',
    getters     => {
      get_foo => 'foo', # 'foo' is the hash key to access
      get_bar => 'bar',
    },
    setters => {
      set_foo => 'foo',
      set_bar => 'bar',
    },
    accessors => {
      foo => 'foo',
      bar => 'bar',
    },
    # "predicates" is an alias for "defined_predicates"
    defined_predicates => {
      defined_foo => 'foo',
      defined_bar => 'bar',
    },
    exists_predicates => {
      has_foo => 'foo',
      has_bar => 'bar',
    },
    lvalue_accessors => { # see below
      baz => 'baz', # ...
    },
    true  => [ 'is_token', 'is_whitespace' ],
    false => [ 'significant' ];
  
  # The imported methods are implemented in fast XS.
  
  # normal class code here.

As of version 1.05, some alternative syntax forms are available:

  package MyClass;
  
  # Options can be passed as a HASH reference, if preferred,
  # which can also help Perl::Tidy to format the statement correctly.
  use Class::XSAccessor {
     # If the name => key values are always identical,
     # the following shorthand can be used.
     accessors => [ 'foo', 'bar' ],
  };

=head1 DESCRIPTION

Class::XSAccessor implements fast read, write and read/write accessors in XS.
Additionally, it can provide predicates such as C<has_foo()> for testing
whether the attribute C<foo> exists in the object (which is different from
"is defined within the object").
It only works with objects that are implemented as ordinary hashes.
L<Class::XSAccessor::Array> implements the same interface for objects
that use arrays for their internal representation.

Since version 0.10, the module can also generate simple constructors
(implemented in XS). Simply supply the
C<constructor =E<gt> 'constructor_name'> option or the
C<constructors =E<gt> ['new', 'create', 'spawn']> option.
These constructors do the equivalent of the following Perl code:

  sub new {
    my $class = shift;
    return bless { @_ }, ref($class)||$class;
  }

That means they can be called on objects and classes but will not
clone objects entirely. Parameters to C<new()> are added to the
object.

The XS accessor methods are between 3 and 4 times faster than typical
pure-Perl accessors in some simple benchmarking.
The lower factor applies to the potentially slightly obscure
C<sub set_foo_pp {$_[0]-E<gt>{foo} = $_[1]}>, so if you usually
write clear code, a factor of 3.5 speed-up is a good estimate.
If in doubt, do your own benchmarking!

The method names may be fully qualified. The example in the synopsis could
have been written as C<MyClass::get_foo> instead
of C<get_foo>. This way, methods can be installed in classes other
than the current class. See also: the C<class> option below.

By default, the setters return the new value that was set,
and the accessors (mutators) do the same. This behaviour can be changed
with the C<chained> option - see below. The predicates return a boolean.

Since version 1.01, C<Class::XSAccessor> can generate extremely simple methods which
just return true or false (and always do so). If that seems like a
really superfluous thing to you, then consider a large class hierarchy
with interfaces such as L<PPI>. These methods are provided by the C<true>
and C<false> options - see the synopsis.

C<defined_predicates> check whether a given object attribute is defined.
C<predicates> is an alias for C<defined_predicates> for compatibility with
older versions of C<Class::XSAccessor>. C<exists_predicates> checks
whether the given attribute exists in the object using C<exists>.

=head1 OPTIONS

In addition to specifying the types and names of accessors, additional options
can be supplied which modify behaviour. The options are specified as key/value pairs
in the same manner as the accessor declaration. For example:

  use Class::XSAccessor
    getters => {
      get_foo => 'foo',
    },
    replace => 1;

The list of available options is:

=head2 replace

Set this to a true value to prevent C<Class::XSAccessor> from
complaining about replacing existing subroutines.

=head2 chained

Set this to a true value to change the return value of setters
and mutators (when called with an argument).
If C<chained> is enabled, the setters and accessors/mutators will
return the object. Mutators called without an argument still
return the value of the associated attribute.

As with the other options, C<chained> affects all methods generated
in the same C<use Class::XSAccessor ...> statement.

=head2 class

By default, the accessors are generated in the calling class. The
the C<class> option allows the target class to be specified.

=head1 LVALUES

Support for lvalue accessors via the keyword C<lvalue_accessors>
was added in version 1.08. At this point, B<THEY ARE CONSIDERED HIGHLY
EXPERIMENTAL>. Furthermore, their performance hasn't been benchmarked
yet.

The following example demonstrates an lvalue accessor:

  package Address;
  use Class::XSAccessor
    constructor => 'new',
    lvalue_accessors => { zip_code => 'zip' };
  
  package main;
  my $address = Address->new(zip => 2);
  print $address->zip_code, "\n"; # prints 2
  $address->zip_code = 76135; # <--- This is it!
  print $address->zip_code, "\n"; # prints 76135

=head1 CAVEATS

Probably won't work for objects based on I<tied> hashes. But that's a strange thing to do anyway.

Scary code exploiting strange XS features.

If you think writing an accessor in XS should be a laughably simple exercise, then
please contemplate how you could instantiate a new XS accessor for a new hash key
that's only known at run-time. Note that compiling C code at run-time a la L<Inline::C|Inline::C>
is a no go.

Threading. With version 1.00, a memory leak has been B<fixed>. Previously, a small amount of
memory would leak if C<Class::XSAccessor>-based classes were loaded in a subthread without having
been loaded in the "main" thread. If the subthread then terminated, a hash key and an int per
associated method used to be lost. Note that this mattered only if classes were B<only> loaded
in a sort of throw-away thread.

In the new implementation, as of 1.00, the memory will still not be released, in the same situation,
but it will be recycled when the same class, or a similar class, is loaded again in B<any> thread.

=head1 SEE ALSO

=over

=item * L<Class::XSAccessor::Array>

=item * L<AutoXS>

=back

=head1 AUTHORS

=over

=item * Steffen Mueller E<lt>smueller@cpan.orgE<gt>

=item * chocolateboy E<lt>chocolate@cpan.orgE<gt>

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2020 by Steffen Mueller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
